### ISSUE TITLE
Type of "xstate" could not be determined because it refers to itself

### ISSUE THREAD

**Issue by Dr-Irv** [microsoft/pyright#4513]
**Describe the bug**
I am getting an intermittent bug in the pyright bundled with pylance

**To Reproduce**
Too intermittent!

**Expected behavior**
No message

**Screenshots or Code**
```python
def myfunc(xstate: Optional[AppState] = None):
    if xstate is not None:
        ystate: AppState = cast(AppState, xstate)
```

I get the message "Type of "xstate" could not be determined because it refers to itself" and `xstate` is highlighted.

But here is the oddity.  The simple example above does not report the error.  The error is in a function with 5 parameters.

If I do some edits to the function by commenting out code, and then undo those edits, the error goes away.  But if I edit another module, it might reappear.

It does not occur in the command line pyright.  

I have no idea how to provide a reproducible error because commenting out the function contents, and then uncommenting them makes the error go away.

**VS Code extension or command-line**
VSCode version of pylance is v2023.1.31
Command line pyright is 1.1.290

**Additional context**
I believe I occasionally saw the error with version 2023.1.21

**Comment by erictraut** [comment]
I'll need you to provide a sample that repros the problem. Can you provide a more complete sample?

**Comment by hmc-cs-mdrissi** [comment]
I've seen this error as well with pylance but also have trouble finding any MVP/steps to reproduce it. Usually when it happens I just close vscode and re-open it and it fixes itself. I'd estimate it happens a couple times a week for me and has happened on variety of different files. I think it started about a month or two back. Trickiest thing is it usually happens after a while of usage so I'm guessing it's some specific sequence of files/cached state and I don't think I've ever had it trigger on first file I open in a session.

**Comment by erictraut** [comment]
This sounds like an evaluation ordering bug. Pyright is a lazy type evaluator, so you can point to any identifier in a program, and it will recursively evaluate only the expressions that are required to determine the type of the requested identifier. It caches the type of all expressions it encounters along the way. In theory, you can ask it to evaluate the type of identifiers in any order, and it should result in exactly the same types for all expressions. However, if there is a bug, it's possible that evaluating identifiers in one order could produce different results from another order.

The way that I normally try to repro problems like this is by making a small modification to the file (e.g. by typing a space on the last line) and immediately jiggling the mouse pointer over a particular identifier. The modification causes pyright to discard all cached information, and hovering the mouse over an identifier queues a "hover text" request that will cause pyright to evaluate the type of the identifier first, before it evaluates anything else in the file. You need to be really quick to queue the hover text request first. If you miss it, the types of all identifiers in the file will be evaluated from top to bottom. I repeat this technique once for each identifier in the suspect function looking for errant diagnostics to pop up. I can typically narrow it down to a particular identifier in a minute or less.

If you're able to use this technique to generate errant diagnostics, please post the code and tell me which identifier appears to be involved.

**Comment by Dr-Irv** [comment]
> If you're able to use this technique to generate errant diagnostics, please post the code and tell me which identifier appears to be involved.

I tried.  Again and again.  It is so intermittent.  I happened to reformat all my source files with `black`, and now cannot reproduce it at all.

One thing I tried was to duplicate the file (`manage.py`) and rename it to `manage2.py`.   Then change the one reference that imports `manage` to `manage2`.  The error doesn't appear.  Change it back and it does.  But that was all before I did the reformatting with `black`.

I can't spend this much time trying to make a reproducible sample, because anything I do to try to reproduce it once, doesn't happen a second or third time.

**Comment by erictraut** [comment]
OK, thanks for trying to repro. 

Without a repro case, there's nothing I can do to further diagnose the problem, so I'm going to close the issue. If you find a way to repro it, I'm happy to reopen and look at it.

**Comment by Dr-Irv** [comment]
> OK, thanks for trying to repro.
> 
> Without a repro case, there's nothing I can do to further diagnose the problem, so I'm going to close the issue. If you find a way to repro it, I'm happy to reopen and look at it.

Is there any debugging output that could be turned on, i.e., that shows the types as they are being evaluated?

**Comment by CJTurpie** [comment]
I'm also experiencing this issue regularly (multiple times a day). Is there any logging or diagnostics information that I could provide when this happens to help diagnose the issue?

**Comment by kaigu1997** [comment]
I have also met the same issue, the min reproducible code is following
```
import gpytorch
import torch


class MyGPR(gpytorch.models.ExactGP):
	def __init__(self, x: torch.Tensor, y: torch.Tensor, likelihood: gpytorch.likelihoods.Likelihood, kernel: gpytorch.kernels.Kernel):
		super().__init__(x, y, likelihood)
		self.mean: gpytorch.means.Mean = gpytorch.means.ZeroMean()
		self.cov: gpytorch.kernels.Kernel = kernel

	def forward(self, x: torch.Tensor) -> gpytorch.distributions.MultivariateNormal:
		Mean = self.mean(x)
		assert isinstance(Mean, torch.Tensor)
		return gpytorch.distributions.MultivariateNormal(Mean, self.cov(x))


def gpytorch_train(
	x: torch.Tensor,
	y: torch.Tensor,
	kernel: gpytorch.kernels.Kernel
) -> MyGPR:
	model: MyGPR = MyGPR(x, y, gpytorch.likelihoods.FixedNoiseGaussianLikelihood(torch.full(x.shape[:-1], gpytorch.settings.min_fixed_noise.value(torch.double) or 1e-8)).train(), kernel).train() # cannot simplified to 1e-8
	loss: torch.Tensor = torch.ones(1)
	hessian: torch.Tensor = torch.eye(1) # cannot move into for loop
	for _ in range(1): # cannot delete this
		for param in model.parameters():
			param.grad = torch.autograd.grad(loss, param, None, True, True, True, True, False, True)[0] # cannot change to ones_like(param)
			if torch.any(torch.isnan(param.grad)).item():
				return model # cannot delete this
		if torch.any(torch.isnan(hessian)).item():
			return model # cannot delete this
		hessian = (hessian + hessian.T) / 2.0 # cannot delete this
		param_combined: torch.Tensor = torch.cat([param.data.reshape(-1) for param in model.parameters()])
		loss = param_combined.norm() ### Type of "param_combined" could not be determined because it refers to itself
	return model.eval()
```
The error appears on line 34, variable `param_combined`. Change `norm` to `sum` or `prod` and the like would have the same error info. Many of the context, once been deleted or modified, would cause the error to disappear, as I have denoted in comments.

VSCode version: 1.93.1 (system setup)
Submit: 38c31bc77e0dd6ae88a4e9cc93428cc27a56ba40
Date: 2024-09-11T17:20:05.685Z
Electron: 30.4.0
ElectronBuildld: 10073054
Chromium: 124.0.6367.243
Nodejs: 20.15.1
V8: 12.4.254.20-electron.0
OS: Windows NT x64 10.0.19045 (in fact this is on WSL2, Ubuntu 24.04.1)

Python: 3.12.3
VSCode Python extension: v2024.14.1
Pylance extension: v2024.9.1

Hope this helps.

**Comment by erictraut** [comment]
@kaigu1997, thanks for the code sample! That was the repro that I needed to find and fix the problem. This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.381.
